lib/Runtime/Library/InJavascript/Intl.js

```diff
@@ -552,62 +562,167 @@
      *
-     * @param {String} langtag a candidate BCP47 langtag
+     * Note that ICU does not implement this correctly for our usage because it is
+     * extremely permissive about what it will allow -- completely invalid language tags can
+     * pass through a round of uloc_forLanguageTag/uloc_toLanguageTag or uloc_canonicalize
+     * even if they are completely bogus.
+     *
+     * ECMA402: #sec-isstructurallyvalidlanguagetag
+     *
+     * @param {String} locale The locale to check
+     * @returns {Boolean}
      */
-    const parseLangtag = function (langtag) {
-        const parts = _.match(langtag, LANG_TAG_RE);
-        if (!parts || !parts[1]) {
-            return null;
+    const IsStructurallyValidLanguageTag = function (locale) {
+        const parsed = parseLangtag(locale);
+        if (parsed === null) {
+            return false;
         }
 
-        const ret = { language: parts[1], base: parts[1] };
-        if (parts[2]) {
-            ret.script = parts[2];
-            ret.base += "-" + parts[2];
-        }
+        // check duplicate variants
+        if (parsed.variants) {
+            const variants = _.split(parsed.variants, "-");
+            const reducedVariants = _.reduce(variants, function (reduced, variant) {
+                variant = _.toLowerCase(variant);
+                if (_.arrayIndexOf(reduced, variant) === -1) {
+                    _.push(reduced, variant);
+                }
 
-        if (parts[3]) {
-            ret.region = parts[3];
-            ret.base += "-" + parts[3];
-        }
+                return reduced;
+            }, []);
 
-        if (parts[4]) {
-            // strip the first character to turn -u-... into u-...
-            ret.extension = _.substring(parts[4], 1);
+            if (variants.length !== reducedVariants.length) {
+                return false;
+            }
         }
 
-        return ret;
-    }
+        if (parsed.extensions) {
+            const singletons =
+                _.map(
+                    _.filter(
+                        _.split(parsed.extensions, "-"),
+                        (element) => element.length === 1
+                    ),
+                    (element) => _.toLowerCase(element)
+                );
 
-    function IsWellFormedCurrencyCode(code) {
-        code = Internal.ToString(code);
+            const seen = _.fill(new platform.Array(26), false);
+            const lowerOffset = _.charCodeAt("a", 0);
+            const upperOffset = _.charCodeAt("A", 0);
 
-        if (!CURRENCY_CODE_RE) {
-            InitializeCurrencyRegExp();
+            const reducedSingletons = _.reduce(singletons, function (reduced, singleton) {
+                const singletonCode = _.charCodeAt(singleton, 0);
+
+                // this is basically String.prototype.toLowerCase when we know we are only dealing with alpha ASCII
+                const el = singletonCode >= upperOffset ? singletonCode - upperOffset : singletonCode - lowerOffset;
```

Comment:
This logic looks wrong, since 'A' is lower than 'a', so the false branch of this conditional will result in negative values.

Reply:
(And the true branch mishandles lowercase characters)